Server/ReasnAPI/ReasnAPI/Services/ImageService.cs

```diff
@@ -1,16 +1,101 @@
+using Image = ReasnAPI.Models.Database.Image;
+
+namespace ReasnAPI.Services;
+public class ImageService (ReasnContext context) 
+{
+    public ImageDto CreateImage(ImageDto imageDto)
+    {
+        var image = context.Images.FirstOrDefault(r => r.ObjectId == imageDto.ObjectId && r.ObjectTypeId == imageDto.ObjectTypeId);
+        if (image != null)
+        {
+            return null;
+        }
+
+        var newImage = new Image
+        {
+            ImageData = imageDto.ImageData,
+            ObjectId = imageDto.ObjectId,
+            ObjectTypeId = imageDto.ObjectTypeId
+        };
+
+        context.Images.Add(newImage);
+        context.SaveChanges();
+        return imageDto;
+    }
+
+    public ImageDto UpdateImage(int imageId,ImageDto imageDto)
+    {
+        var image = context.Images.FirstOrDefault(r => r.Id == imageId);
+        if (image == null)
+        {
+            return null;
+        }
+
+        image.ObjectId = imageDto.ObjectId;
+        image.ImageData = imageDto.ImageData;
+        image.ObjectTypeId = imageDto.ObjectTypeId;
+
+        context.Images.Update(image);
+        context.SaveChanges();
+        return imageDto;
+    }
+
+    public void DeleteImage(int id)
+    {
+        var image = context.Images.FirstOrDefault(r => r.Id == id);
+        if (image == null) { return;}
+        context.Images.Remove(image);
+        context.SaveChanges();
+     
+    }
+
+    public ImageDto GetImageById(int id)
+    {
+        var image = context.Images.FirstOrDefault(r => r.Id == id);
+        if (image == null)
+        {
+            return null;
+        }
+
+        var imageDto = new ImageDto
+        {
+            ImageData = image.ImageData,
+            ObjectId = image.ObjectId,
+            ObjectTypeId = image.ObjectTypeId
+        };
+
+        return imageDto;
+    }
 
-namespace ReasnAPI.Services {
-    public class ImageService (ReasnContext context) {
-        private readonly ReasnContext _context = context;
+    public IEnumerable<ImageDto> GetAllImages()
+    {
+        var images = context.Images.ToList();
+        
+        var imageDtos = images.Select(image => new ImageDto
+        {
+            ImageData = image.ImageData,
+            ObjectId = image.ObjectId,
+            ObjectTypeId = image.ObjectTypeId
+        }).ToList();
+
+        return imageDtos;
+    }
 
-        /* TODO: Create following functions for this class
-         * create
-         * update
-         * delete
-         * get by ID
-         * get list by filter
-         * get all
-         */
+    public IEnumerable<ImageDto> GetImagesByFilter(Expression<Func<Image, bool>> filter)
+    {
+        var images = context.Images.Where(filter).ToList();
+
+        var imageDtos = images.Select(image => new ImageDto
+        {
+            ImageData = image.ImageData,
+            ObjectId = image.ObjectId,
+            ObjectTypeId = image.ObjectTypeId
+        }).ToList();
+
+        return imageDtos;

```

Comment:
As far i know if the images collection is empty, the Select method will not iterate over any elements, and then, no ImageDto objects will be created. It will return an empty IEnumerable<ImageDto>.  I might be wrong with this one.

Reply:
So please check documentation what Where returns when there is no rows satisfying the condition, but I am pretty sure you are right